### PR TITLE
ng_ipv6: initial import

### PR DESCRIPTION
**Pull request by authmillenon**
Depends on
 * ~~#2285~~ (merged)
 * ~~#2433~~ (merged)
 * ~~#2437~~ (merged)
 * ~~#2449~~ (merged)
 * ~~#2455~~ (merged)
 * ~~#2461~~ (merged)
 * ~~#2546~~ (merged)
 * ~~#2553~~ (merged)
 * ~~#2575~~ (merged)
 * ~~#2600~~ (merged)
 * ~~#2731~~ (merged)
 * ~~#2735~~ (merged)

### CODE REVIEW

**Comment by authmillenon** [comment]
Rebased to master and depending PRs

**Comment by authmillenon** [comment]
Rebased to current master and the depending PRs

**Comment by authmillenon** [comment]
[ng_ipv6.h line 14 and onwards](https://github.com/authmillenon/RIOT/blob/ng_ipv6/feat/initial/sys/include/net/ng_ipv6.h#L14) contains my idea of how to document NETAPI behavior. @haukepetersen, @jfischer-phytec-iot, @LudwigOrtmann, @Lotterleben, @OlegHahm, @PeterKietzmann, @thomaseichinger what do you think

**Comment by authmillenon** [comment]
No longer WIP, squashed and rebased to master and depending PRs.

**Comment by authmillenon** [comment]
(tests are still missing)

**Comment by authmillenon** [comment]
Rebased to depending PRs and squashed

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by authmillenon** [comment]
Rebased to current dependencies.

**Comment by authmillenon** [comment]
I've reworked the whole thing a little to make the implementation a little clearer. This is why I squashed on rebase.

**Comment by authmillenon** [comment]
Rebased to current master and dependencies

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by authmillenon** [comment]
Rebased to current master.

**Comment by authmillenon** [comment]
All dependencies were merged.

**Comment by authmillenon** [comment]
Rebased to current master.

**Comment by OlegHahm** [comment]
> All dependencies were merged.

Awesome - so IPv6 support is only one step away? :grinning:

**Comment by authmillenon** [comment]
*Basic* IPv6 support, yes. #2555, #2612, and some form of extension header handling are needed too ;-)

**Comment by OlegHahm** [comment]
Who needs NDP? :wink:

**Comment by authmillenon** [comment]
(Oh and would be great if the soft dependency #2804 can be merged beforehand so I can make https://github.com/RIOT-OS/RIOT/pull/2454/files#diff-c0ab51e027bbe0cfdb3482d3e29d6f34R157 safer ;-) ;-))

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by authmillenon** [comment]
Rebased to current master.

File: Makefile.dep [hunk 1]
```diff
@@ -58,6 +58,20 @@ ifneq (,$(filter ng_ipv6_hdr,$(USEMODULE)))
   USEMODULE += ng_pktbuf
 endif
 
+ifneq (,$(filter ng_ipv6_router,$(USEMODULE)))
+  USEMODULE += ng_ipv6
+  USEMODULE += ng_fib
```
Comment (OlegHahm) [inline Makefile.dep]: Why is this dependency? (I understand the conceptional dependency, but I couldn't find the implementational dependency.)

File: sys/include/net/ng_ipv6.h [hunk 2]
```diff
@@ -22,14 +28,66 @@
...
+ */
+#ifndef NG_IPV6_PRIO
+#define NG_IPV6_PRIO            (PRIORITY_MAIN - 3)
+#endif
+
+/**
+ * @brief   Default message queue size to use for the IPv6 thread.
+ */
+#ifndef NG_IPV6_MSG_QUEUE_SIZE
+#define NG_IPV6_MSG_QUEUE_SIZE  (8U)
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6.h]: Is this number based on a certain assumption or just an educated guess?

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+
+#include "net/ng_ipv6/nc.h"
+#include "net/ng_ipv6/netif.h"
+
+#include "net/ng_ipv6.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _MAX_L2_ADDR_LEN    (8U)
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Wouldn't it make more sense to define this outside IP? In a similar fashion how maximum payload is defined for the transceiver module, for instance.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+
+    payload->type = ng_nettype_from_protnum(nh);
+
+    /* TODO: add ICMPv6 and extension header handling */
+
+    receiver_num = ng_netreg_num(payload->type, NG_NETREG_DEMUX_CTX_ALL) +
+                   ng_netreg_num(NG_NETTYPE_IPV6, nh);
+
+    if (receiver_num == 0) {
+        DEBUG("udp: unable to forward packet as no one is interested in it\n");
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: udp?

File: Makefile.dep (hunk 1 above)
Comment (authmillenon) [inline Makefile.dep]: You're right. Since I kept routing out for now (and it will become part of the neighbor discovery anyways), this whole `ifneq` can be removed. Will do.

File: sys/include/net/ng_ipv6.h (hunk 2 above)
Comment (authmillenon) [inline sys/include/net/ng_ipv6.h]: It's basically my go-to number for message queue size. But https://lists.riot-os.org/pipermail/devel/2015-April/002458.html seems to indicate it is a good one in this context.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+static kernel_pid_t _ipv6_pid = KERNEL_PID_UNDEF;
+
+#if ENABLE_DEBUG
+static char addr_str[NG_IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+static void _ipv6_receive(kernel_pid_t iface, ng_pktsnip_t *pkt);
+static void _ipv6_send(ng_pktsnip_t *pkt, bool prep_hdr);
+static void *_ipv6_event_loop(void *args);
+static void _dispatch_ipv6_pkt(ng_nettype_t type, uint32_t demux_ctx,
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Could add a single line of documentation for each function what it is actually doing?

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 3 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: See https://github.com/RIOT-OS/RIOT/pull/2833/files#diff-0010497e859015264b22918ac7e1109fR39

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 5 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Can do.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 4 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Copy-Pasta :)

Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: will fix

File: sys/include/net/ng_ipv6.h (hunk 2 above)
Comment (OlegHahm) [inline sys/include/net/ng_ipv6.h]: I see, thanks for explanation.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 3 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Can we not just use this define here?

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+    if (tmp == NULL) {
+        ng_pktbuf_release(pkt);
+    }
+
+    return tmp;
+}
+
+static void *_ipv6_event_loop(void *args)
+{
+    msg_t msg, reply, msg_q[NG_IPV6_MSG_QUEUE_SIZE];
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: I just realized that this was done for a couple of other modules, too, but I wonder if allocating the message queue on the stack is a good idea. It probably doesn't hurt for the default size (8 * 8 bytes), but may become an issue if someone increases the queue size on a board with aggressively optimized stack sizes.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+    msg_init_queue(msg_q, NG_IPV6_MSG_QUEUE_SIZE);
+
+    me_reg.demux_ctx = NG_NETREG_DEMUX_CTX_ALL;
+    me_reg.pid = thread_getpid();
+
+    /* register interest in all IPv6 packets */
+    ng_netreg_register(NG_NETTYPE_IPV6, &me_reg);
+
+    /* preinitialize ACK */
+    reply.type = NG_NETAPI_MSG_TYPE_ACK;
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Shouldn't this go inside the loop? The receiver may overwrite the type of the message.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 3 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: As soon as #2833 is merged we can. Apart from that it does not really hurt for now here either.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 7 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: It's always an `ACK`. If the receiver changes the type the change happens to the instance of `msg_t` in their stack. The local type field here shouldn't change.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 6 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Not part of this discussion, but IMHO then you always can change the stack size, too.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 7 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Ah, yes, sorry, stupid thinking.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 3 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Oops, I didn't realized that the link was part of a PR.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+        DEBUG("ipv6: Send receive command for %p to %" PRIu16 "\n", (void *)pkt,
+              entry->pid);
+        msg_send(&msg, entry->pid);
+        entry = ng_netreg_getnext(entry);
+    }
+}
+
+static void _ipv6_receive(kernel_pid_t iface, ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *ipv6, *payload;
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: I find the name `payload` confusing in this context, but that's just a detail.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+        if (!ng_ipv6_hdr_is_ipv6_hdr(ipv6->data)) {
+            DEBUG("ipv6: Received packet was not IPv6, dropping packet\n");
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        payload = _start_write(payload);
+    }
+    else {
+        if (!ng_ipv6_hdr_is_ipv6_hdr(payload->data)) {
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: just as a side note: this function name is somewhat funny.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 6 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: If you know that the queue is allocated on the stack... But you're right will open an issue.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 9 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: I know. That's why I called 6LoWPAN check functions e.g. `ng_sixlowpan_frag_is()`. I can do this for this function too, but I would prefer to do this as a follow-up.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 8 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Why. The variable is the payload of the L2 frame before parsing and then becomes the payload of the IPv6 packet afterwards (while `ipv6` becomes its header).

**Comment by authmillenon** [comment]
Addressed comments so far.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 9 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: 100% agreed that this is out of scope of this PR.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 8 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Because - at the point of commenting - I only saw it used as L2 payload, why a variable called `payload` in the IPv6 implementation file suggests that this refers to L2 payload. Since it is used for both, I think it is somehow fine.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 6 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: I did: #2851

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 9 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: However, you could move this check before https://github.com/RIOT-OS/RIOT/pull/2454/files#diff-c0ab51e027bbe0cfdb3482d3e29d6f34R433 to exit early.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+    _dispatch_rcv_pkt(NG_NETTYPE_IPV6, nh, pkt);
+}
+
+/* internal functions */
+static inline ng_pktsnip_t *_start_write(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *tmp = ng_pktbuf_start_write(pkt);
+
+    if (tmp == NULL) {
+        ng_pktbuf_release(pkt);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Why is this released here again?

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 9 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: I can't I don't know at that point if `payload` is an IPv6 header or it's payload

Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Hm, ok.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+            DEBUG("ipv6: Received packet was not IPv6, dropping packet\n");
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        pkt = ng_pktbuf_start_write(pkt);
+
+        if (pkt == NULL) {
+            DEBUG("ipv6: unable to get write access to packet, drop it\n");
+            ng_pktbuf_release(pkt);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Calling release on a NULL pointer?

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+            }
+
+            ng_pktbuf_release(ipv6->next);  /* remove headers around IPV6 */
+            ipv6->next = pkt;           /* reorder for sending */
+            pkt->next = NULL;
+            _send(ipv6, false);
+        }
+        else {
+            DEBUG("ipv6: hop limit reached 0: drop packet\n");
+            ng_pktbuf_release(pkt);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: return is missing

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 11 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Now I remember, why I introduced `_start_write()` in the first place :(

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 12 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Fixed

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+
+        switch (msg.type) {
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("ipv6: NG_NETAPI_MSG_TYPE_RCV received\n");
+                _receive((ng_pktsnip_t *)msg.content.ptr);
+                break;
+
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("ipv6: NG_NETAPI_MSG_TYPE_SND received\n");
+                _send((ng_pktsnip_t *)msg.content.ptr, true);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: What exactly is the meaning of the second argument?

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+                return;
+            }
+
+            LL_PREPEND(pkt, netif);
+
+            DEBUG("ipv6: send multicast over interface %" PRIkernel_pid "\n", ifs[i]);
+            /* mark as multicast */
+            ((ng_netif_hdr_t *)netif->data)->flags |= NG_NETIF_HDR_FLAGS_MULTICAST;
+            /* and send to interface */
+            ng_netapi_send(iface, pkt);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Code duplication with l312--316. Please inline.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 13 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: That the IPv6 header should be changed if need be (setting of source address, next header, and length field and upper layer checksum resulting from that in particular) if `true`. You don't want this (`false`) if you just redirect the packet.

Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: (see https://github.com/RIOT-OS/RIOT/pull/2454/files#diff-c0ab51e027bbe0cfdb3482d3e29d6f34R320) for documentation. Will move it to the top.

**Comment by authmillenon** [comment]
Addressed more comments

**Comment by OlegHahm** [comment]
ACK. Please squash.

**Comment by authmillenon** [comment]
Squashed

**Comment by authmillenon** [comment]
Applied the following bugfix and squashed immediately:

```diff
diff --git a/sys/net/network_layer/ng_ipv6/ng_ipv6.c b/sys/net/network_layer/ng_ipv6/ng_ipv6.c
index 1be5142..b53d5d3 100644
--- a/sys/net/network_layer/ng_ipv6/ng_ipv6.c
+++ b/sys/net/network_layer/ng_ipv6/ng_ipv6.c
@@ -236,7 +236,8 @@ static int _fill_ipv6_hdr(kernel_pid_t iface, ng_pktsnip_t *ipv6,
     return 0;
 }
 
-static inline void _send_multicast_over_iface(kernel_pid_t iface, ng_pktsnip_t *pkt)
+static inline void _send_multicast_over_iface(kernel_pid_t iface, ng_pktsnip_t *pkt,
+                                              ng_pktsnip_t *netif)
 {
     DEBUG("ipv6: send multicast over interface %" PRIkernel_pid "\n", ifs[i]);
     /* mark as multicast */
@@ -299,7 +300,7 @@ static void _send_multicast(kernel_pid_t iface, ng_pktsnip_t *pkt,
 
             LL_PREPEND(pkt, netif);
 
-            _send_multicast_over_iface(iface, pkt);
+            _send_multicast_over_iface(iface, pkt, netif);
         }
     }
     else {
@@ -312,7 +313,7 @@ static void _send_multicast(kernel_pid_t iface, ng_pktsnip_t *pkt,
             }
         }
 
-        _send_multicast_over_iface(iface, pkt);
+        _send_multicast_over_iface(iface, pkt, netif);
     }
 }
```

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 9 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: See #2857.
